### PR TITLE
Move interpolate key and keys from core to common so it can be reused by the upcoming test engine.

### PR DESCRIPTION
**Pull request by nolag**
See use in https://github.com/smartcontractkit/chainlink/pull/14248

### CODE REVIEW

File: pkg/workflows/execution/interpolation.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
+package execution
```
Comment (jmank88) [inline pkg/workflows/execution/interpolation.go:1]: Maybe consider `exec`?
```suggestion
package exec
```
https://go.dev/blog/package-names

File: pkg/workflows/execution/interpolation.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+			inner, ok := v[r]
+			if !ok {
+				return "", fmt.Errorf("could not find ref part `%s` (ref: `%s`) in `%+v`", r, key, v)
+			}
+
+			val = inner
+		case []any:
+			i, err := strconv.Atoi(r)
+			if err != nil {
+				return "", fmt.Errorf("could not interpolate ref part `%s` (ref: `%s`) in `%+v`: `%s` is not convertible to an int", r, key, v, r)
```
Comment (jmank88) [inline pkg/workflows/execution/interpolation.go:67]: Should `err` be included?

File: pkg/workflows/execution/results.go [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package execution
+
+import "github.com/smartcontractkit/chainlink-common/pkg/values"
+
+type Results interface {
+	GetResultForStep(string) (*Result, bool)
```
Comment (jmank88) [inline pkg/workflows/execution/results.go:6]: Does the `Get` prefix add value here?
```suggestion
	ResultForStep(string) (*Result, bool)
```
https://go.dev/doc/effective_go#Getters

File: pkg/workflows/execution/interpolation.go (hunk 2 above)
Comment (nolag) [reply pkg/workflows/execution/interpolation.go:67]: was copied, but will add

**Review by jmank88 — APPROVED** [review]

File: pkg/workflows/exec/interpolation.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+// it with its corresponding value in `state`.
+//
+// A key is valid if it contains at least two parts, with:
+//   - the first part being the workflow step's `ref` variable
+//   - the second part being one of `inputs` or `outputs`
+//
+// If a key has more than two parts, then we traverse the parts
+// to find the value we want to replace.
+// We support traversing both nested maps and lists and any combination of the two.
+func InterpolateKey(key string, state Results) (any, error) {
```
Comment (cedric-cordenier) [inline pkg/workflows/exec/interpolation.go:22]: Do we need to use an interface here? Under the hood this is just a map of State objects isn't it? Why not move that out too -- that feels like it simplifies the implementation

Comment (jmank88) [reply pkg/workflows/exec/interpolation.go:22]: If that is not possible, another in-between approach could be to accept a func here:
```suggestion
func InterpolateKey(key string, resultForStep func(string) (*Result, bool)) (any, error) {
```
Then caller just passes a reference to its method: `exec.InterpolateKey(key, results.ResultsForStep)

**Review by cedric-cordenier — APPROVED** [review]

**Pull request merged**